solvation_analysis/analysis.py

```diff
@@ -0,0 +1,144 @@
+        # if not kernel:
+        #     self.kernel = identify_solvation_cutoff
+        self.kernel_kwargs = {} if kernel_kwargs is None else kernel_kwargs
+        self.rdf_kwargs = {"range": (0, 8.0)} if rdf_kwargs is None else rdf_kwargs
+        # if not kernel_kwargs:
+        #     self.kernel_kwargs = {}
+
+        self.solute = solute
+        self.solvents = solvents
+        self.u = self.solute.universe
+        self.rdf_plots = {}
+        self.rdf_data = {}
+        self.ion_speciation = None
+        self.ion_pairing = None
+        self.coordination_numbers = None
+        self.solvation_frames = []
+
+    @classmethod
+    def _plot_solvation_radius(cls, bins, data, radius):
+        fig, ax = plt.subplots()
+        ax.plot(bins, data, "b-", label="rdf")
+        ax.axvline(radius, color="r", label="solvation radius")
+        ax.set_xlabel("Radial Distance (A)")
+        ax.set_ylabel("Probability Density")
+        ax.legend()
+        return fig, ax
+
+    def _prepare(self):
+        for name, solvent in self.solvents.items():
+            # generate and save RDFs
+            rdf = InterRDF(self.solute, solvent, **self.rdf_kwargs)
+            # TODO: specify start and stop args
+            rdf.run()
+            bins, data = rdf.results.bins, rdf.results.rdf
+            self.rdf_data[name] = (data, bins)
+            # generate and save plots
+            if name not in self.radii.keys():
+                self.radii[name] = self.kernel(bins, data, **self.kernel_kwargs)
+            fig, ax = self._plot_solvation_radius(bins, data, self.radii[name])
+            ax.set_title(f"Solvation distance of {name}")
+            self.rdf_plots[name] = fig, ax
+        assert self.solvents.keys() == self.radii.keys(), "Radii missing."
+
+    def _single_frame(self):
+        # initialize empty arrays
+        all_pairs_list = []
+        all_dist_list = []
+        all_tags_list = []
+        for name, solvent in self.solvents.items():
+            pairs, dist = capped_distance(
+                self.solute.positions,
+                solvent.positions,
+                self.radii[name],
+                box=self.u.dimensions,
+            )
+            # replace local ids with absolute ids
+            pairs[:, 1] = solvent.ids[[pairs[:, 1]]]  # TODO: ids vs ix?
+            # extend
+            all_pairs_list.append(pairs)
+            all_dist_list.append(dist)
+            all_tags_list.append(np.full(len(dist), name))  # creating a name array
+        all_pairs = np.concatenate(all_pairs_list)
+        all_dist = np.concatenate(all_dist_list)
+        all_tags = np.concatenate(all_tags_list)
+        # put the data into a data frame
+        all_resid = self.u.atoms[all_pairs[:, 1]].resids
+        solvation_data_np = np.column_stack(
+            (all_pairs[:, 0], all_pairs[:, 1], all_dist, all_tags, all_resid)
+        )
+        solvation_data_pd = pd.DataFrame(
```

Comment:
As I mentioned, I'm really struggling with the best core data structure for this. Right now I am using a list of dataframes, where there is a separate DataFrame for each timestep that is analyzed. Under this organization, I don't think there is a way to avoid calling pd.DataFrame at every step or a great reason to move the logic to conclude. With a different data structure, perhaps that would change.

A few of the issues that make this challenging:

1. the analysis runs over several time steps and at each time step there are many solute molecules
2. the solvent molecules associated with each solute change often
3. analysis is easiest to write on a single time step at a time, so partitioning the different time steps into a list is natural.

Reply:
Did you use `%timeit` in an interactive shell to compare the time to create a typical dataframe vs the corresponding np.array? This would give you an idea how expensive the nicer data structure is and if this is a trade-off worth doing.

Are the per-step dataframes of different size? Do you have it documented somewhere what they should contain, e.g., a wiki page or doc string?

Is this data structure user facing or internal?

Ultimately, if this works for right now, I'd use it and add a comment that this may later change. Right now the goal is to ship a prototype. 🚢 

["Real artists ship!"](https://www.folklore.org/StoryView.py?story=Real_Artists_Ship.txt) — Steve Jobs

Reply:
Here is a quick sketch of my thinking. The consistent indices are the solute and the time step. At a specific solute and time step there are several atoms, which each have an atom_id, a residue_name, and a residue_id. 

I'd be great to have some selection that lets the user select by time step or by solute number and return a series of solvent shells. The challenge is that the atoms in each shell are not consistent.

Maybe would xarray provide a convenient way of breaking up this data?

![image](https://user-images.githubusercontent.com/27712051/128439438-ff04d08a-4991-46dd-b40c-6d32d8bbe70e.png)

Reply:
Oops, missed your questions in my last comment.

> Did you use %timeit in an interactive shell to compare the time to create a typical dataframe vs the corresponding np.array?

I have not. I would have trouble writing most of my analysis with a raw numpy array. I think the pandas is important.

> Are the per-step dataframes of different size? Do you have it documented somewhere what they should contain, e.g., a wiki page or doc string?

Yes they are, the number of solutes stays constant but the number of solvents changes at each frame. I have not yet documented this because I still might change the core data structure. I am chatting with @hmacdope and @richardjgowers on Tuesday to discuss.

Reply:
As you have correctly pointed out @orioncohen, the challenge here is that the number of solvent shell atoms is not consistent and so any resulting numpy arrays would have different dims along the atom axis. This is why squeezing into a dataframe is a good idea on many levels.

I think I need to have a play with this branch before I can say for sure on Tuesday, but perhaps an array or list of `mda.Atomgroups` is a possibility? This thinking stems from the idea that the columns of your dataframe in  the above picture are all atomgroup **attributes**. Eg. solute.resid and solute.ix and solute.resnames.

This means you could possibly lose a whole feature axis, eg in your diagram you can just make `solvent atoms` an atomgroup and have any other attributes queryable. The use of atomgroup also solves the variable axis problem. 

Just a thought, and it may not be quicker as querying a `pd.DataFrame` is quite fast. Perhaps just having an idea of how long it currently takes using a `DataFrame` would be a help.  Lets make this a todo for Tuesday?

Also as @orbeckst emphasised, if it works as it is, then we can refine down the track! Better to have full functionality and then polish. 💯

Reply:
I should also say that we can make a distinction between things that should be available to a user and to a developer.

A developer can be expected to be able to use a private attribute provided it is well documented, while a user is not. At least thats how I see it, others may disagree.

This is all kind of a moot point anyway as nothing is truly "private" in python. If a user wanted they could access `Solution._data` just as easily as `Solution.data` (double underscore name mangling is also a thing but lets not get into that).  Its more a question of documentation and convention.

Reply:
I have elected to put all of the data in the same `Pd.DataFrame` and percolated the changes through all of the analysis functions. This has substantially simplified things and made writing analysis much easier. Thank you for the suggestion @hmacdope.

By putting everything in a single frame, the logic in SolvationData became irrelevant, so I decided to cut it to simplify the package.

The new dataframe contains the same information as above but it is 'tidy'. The frame is now an index in the gigantic dataframe.

Pretty happy with the new solution!

Reply:
Awesome! As @orbeckst stated "simple is best" is almost always the the way.

Reply:
Thanks for the discussion y'all. I am going to resolve this now.